pkg/lib/authenticationflow/authflowclient/client.go

```diff
@@ -0,0 +1,145 @@
+package authflowclient
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/authgear/authgear-server/pkg/lib/config"
+	"github.com/authgear/authgear-server/pkg/util/httputil"
+)
+
+var httpClient = &http.Client{}
+
+func init() {
+	httpClient.Timeout = 30 * time.Second
+}
+
+type Client struct {
+	Context       context.Context
+	HTTPClient    *http.Client
+	LocalEndpoint *url.URL
+	HTTPHost      httputil.HTTPHost
+}
+
+func NewClient(ctx context.Context, mainListenAddr config.MainListenAddr, httpHost httputil.HTTPHost) *Client {
+	// Always use http because we are going to call ourselves locally.
+	localEndpointString := fmt.Sprintf("http://%v", mainListenAddr)
+	localEndpointURL, err := url.Parse(localEndpointString)
+	if err != nil {
+		panic(err)
+	}
+
+	// Only the port is important, the host is always the loopback address.
+	localEndpointURL.Host = fmt.Sprintf("127.0.0.1:%v", localEndpointURL.Port())
+
+	return &Client{
+		Context:       ctx,
+		HTTPClient:    httpClient,
+		LocalEndpoint: localEndpointURL,
+		HTTPHost:      httpHost,
+	}
+}
```

Comment:
I think calling self might cause a deadlock in the server, similar to https://github.com/authgear/authgear-server/issues/3058:

Assume the db connection pool have a max of 2. And now there are two incoming webapp requests at the same time.

1. Two requests start at the same time, and run util it calls a authflow api.
2. Now, two new authflow requests received, however, as the previous two webapp requests are not finished, the db connection pool is still exhausted and therefore no connections can be used to handle the authflow requests. As a result, they wait indefinitely.
3. Finally the requests should timeout.

And I am not sure how many threads are used to handle the requests, but I guess the thread pool will have the same problem.

Reply:
I checked the code again and seems we are only using database connections in authenticationflow.Service, and not using it in the authflow webapp 🤔  However we must be careful in the future to avoid such problem. Maybe we should consider deploying authflow api as a separated service?

Reply:
Authflow-based webapp handlers do not acquire database connection. I used webapp.Session to store authflow state because I do not want to maintain yet another state in the webapp. I think those handlers, and authflowclient can be moved to a new package. The new package is a client implementation of the Authflow API, so database connection is forbidden. Making authflow API a separate service will complicate further the deployment, especially in case of non k8s deployment.

Reply:
ok. And I've also checked the concurrent requests golang http server can handle and it seems a large number, so maybe it is not a problem too. Will you move the packages in this pr too?

Reply:
Moved the authflow client implementation to a new package.